[llvm-objdump] Add inlined function display support

gulfemsavrun opened the pull request:
This patch adds the support for displaying inlined functions into llvm-objdump.

1) It extends the source variable display
support for inlined functions both for ascii and unicode formats.

2) It also introduces a new format called limits-only that only prints a line for the start and end of an inlined function without line-drawing characters.

gulfemsavrun committed:
[llvm-objdump] Add inlined function display support

This patch adds the support for displaying inlined functions
into llvm-objdump.

1) It extends the source variable display
support for inlined functions both for ascii and unicode
formats.

2) It also introduces a new format called line that only
prints a line for the start and end of an inlined function
without line-drawing characters.

jh7370 commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
+        InlinedFuncDie(InlinedFuncDie), Range(Range) {}
+
+  virtual ~InlinedFunction() {}
+  virtual bool liveAtAddress(object::SectionedAddress Addr) override;
```
Here and in various other places, there's no need for `virtual` when the method has `override`.

gulfemsavrun replied:
Done.

jh7370 commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -23,6 +23,62 @@
+          DILineInfoSpecifier::FileLineInfoKind::AbsoluteFilePath, FileName))
+    return;
+
+  assert(!FileName.empty());
```
Are you sure this can never fire? What happens if the debug data genuinely has an empty name for a function, e.g. due to malformed/hand-written debug data?

gulfemsavrun replied:
Ok, I removed the assert and added an early return for such a case.

jh7370 commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -23,6 +23,62 @@
+
+void InlinedFunction::print(raw_ostream &OS, const MCRegisterInfo &MRI) const {
+  const char *MangledCallerName =
+      FuncDie.getName(llvm::DINameKind::LinkageName);
```
Here and in various places throughout, why the `llvm::` prefix?

gulfemsavrun replied:
Done.

jh7370 commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -23,6 +23,62 @@
+      FuncDie.getName(llvm::DINameKind::LinkageName);
+  if (!MangledCallerName)
+    return;
+  std::string CallerName = llvm::demangle(MangledCallerName);
```
I think the call to `demangle` needs to be conditional on the `--demangle` option. Same below. This should also be tested.

gulfemsavrun replied:
I added `Demangle` check.

jh7370 commented on llvm/test/tools/llvm-objdump/X86/debug-inlined-function.s:
```diff
@@ -0,0 +1,632 @@
+# RUN: llvm-objdump %t.o -d --debug-inlined-funcs=line | \
+# RUN:     FileCheck %s --check-prefix=LINE
+
+# RUN: llvm-objdump %t.o -d --debug-inlined-funcs | \
```
You should also have a test for explicitly specifying `unicode`.

gulfemsavrun replied:
I explicitly added `unicode` into the test.

gulfemsavrun replied:
So, I added two tests one with `debug-inlined-funcs` flag and one with `debug-inlined-funcs=unicode`, and the outputs are identical for them. Please let me know if that's not what you requested.

gulfemsavrun replied:
Done.

jh7370 reviewed (commented):
This deserves a release note, I think, including highlighting the renamed indent option.

Also, I think it would be useful to have a test showing the interaction between the variable and inlined function options, so how a variable is displayed inside an inlined function when both are enabled.

On that note, does it even make sense for the two to have different values? E.g. one having unicode and the other ASCII? I wonder if we need to revisit how the option works. I'm not sure what would be the best approach, but perhaps there could be a "format" option and a separate option (or options) to determine what to enable. Alternatively, we just go with it and accept that ASCII + unicode is allowed.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -127,9 +208,9 @@ unsigned LiveVariablePrinter::findFreeColumn() {
-  for (const LiveVariable &LV : LiveVariables) {
-    dbgs() << LV.VarName << " @ " << LV.LocExpr.Range << ": ";
-    LV.print(dbgs(), MRI);
+  for (const auto &LE : LiveElements) {
```
Done.

gulfemsavrun commented on llvm/test/tools/llvm-objdump/X86/debug-inlined-function.s:
```diff
@@ -0,0 +1,632 @@
+## Generated with this compile command, with the source code in Inputs/debug-inlined-functions.c:
+## clang -g -c debug-inlined-function.c -O0 -S -o -
```
This is a typo and thanks for catching that. It should be `-O1`.

gulfemsavrun commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -153,10 +153,16 @@ OPTIONS
   to ``unicode`` if omitted.
 
-.. option:: --debug-vars-indent=<width>
+.. option:: --debug-indent=<width>
```
Done.

gulfemsavrun replied:
I reordered them by ignoring hyphens, so I put `--debuginfod` after all `--debug-*` flags.

gulfemsavrun commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -153,10 +153,16 @@ OPTIONS
+  Distance to indent the source-level variable or inlined function display,
+  relative to the start of the disassembly. Defaults to 52 characters.
+
+.. option:: --debug-inlined-funcs=<format>
```
Done.

gulfemsavrun commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -153,10 +153,16 @@ OPTIONS
+.. option:: --debug-inlined-funcs=<format>
+
+  Print the locations of inlined functions alongside disassembly.
+  ``format`` may be ``unicode``, ``ascii`` or ``line`` defaulting to
```
Done.

gulfemsavrun commented:
Thanks for all the feedback @jh7370! 
> This deserves a release note, I think, including highlighting the renamed indent option.

I added this into the release notes.

> Also, I think it would be useful to have a test showing the interaction between the variable and inlined function options, so how a variable is displayed inside an inlined function when both are enabled.

I extended the case to show how it looks like when `--debug-vars` and `--debug-inlined-functions` options are enabled at the same time.

> On that note, does it even make sense for the two to have different values? E.g. one having unicode and the other ASCII? I wonder if we need to revisit how the option works. I'm not sure what would be the best approach, but perhaps there could be a "format" option and a separate option (or options) to determine what to enable. Alternatively, we just go with it and accept that ASCII + unicode is allowed.

I think we can add a new option like `--debug-format` for the format. It can be shared  between `--debug-vars` and `--debug--inlined-funcs` like `--debug-indent` option. What are you thoughts on the `line` format that I added? This was a feature requested by some of our project developers who do not want to look at line-drawing characters.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -150,44 +231,51 @@ void LiveVariablePrinter::addCompileUnit(DWARFDie D) {
-    ActiveCols[ColIdx].LiveOut = LV.liveAtAddress(NextAddr);
+    }
+    CheckedElementIdxs.insert(ActiveCols[ColIdx].ElementIdx);
+    auto &LE = LiveElements[ActiveCols[ColIdx].ElementIdx];
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -336,6 +425,24 @@ void LiveVariablePrinter::printAfterInst(formatted_raw_ostream &OS) {
+  // Print a line to idenfity the start of an inlined function if line format
+  // is specified.
+  if (DbgInlinedFunctions == DVLine)
+    for (const auto &LE : LiveElements)
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
+      : LiveElement(FunctionName, Unit, FuncDie),
+        InlinedFuncDie(InlinedFuncDie), Range(Range) {}
+
+  virtual ~InlinedFunction() {}
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -150,44 +231,51 @@ void LiveVariablePrinter::addCompileUnit(DWARFDie D) {
+  SmallSet<unsigned, 8> CheckedElementIdxs;
   for (unsigned ColIdx = 0, End = ActiveCols.size(); ColIdx < End; ++ColIdx) {
-    if (!ActiveCols[ColIdx].isActive())
+    if (!ActiveCols[ColIdx].isActive()) {
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -276,17 +364,18 @@ void LiveVariablePrinter::printBetweenInsts(formatted_raw_ostream &OS,
           OS << "  ";
       }
 
+      auto &LV = LiveElements[ActiveCols[ColIdx].ElementIdx];
```
Done.

jh7370 reviewed (commented):
Please don't force push to update PR branches once they have active reviewers, in line with the LLVM GitHub policy (see https://llvm.org/docs/GitHub.html#updating-pull-requests and the following section): it impairs the reviewing experience, making it harder and slower for me to review your changes. Also, please don't resolve conversations that I initiated, as I use the "resolved" status to track whether I'm happy with the response you've made (it's not unusual for people to miss/misunderstand/etc comments and do the wrong thing). I've gone into more detail on Discourse in the past on this topic.

> I think we can add a new option like --debug-format for the format. It can be shared between --debug-vars and --debug--inlined-funcs like --debug-indent option.

Having given it a bit more thought, I don't think we want an extra option and it's fine to leave the behaviour as you've implemented. If users want to specify different formats for the two, that's on them.

> What are your thoughts on the line format that I added? This was a feature requested by some of our project developers who do not want to look at line-drawing characters.

I think the format looks reasonable. I'm not certain about the naming though, as `line` doesn't really convey anything to me. I'm thinking something more like `limits-only` or something like that, but am also open to other ideas.

@MaskRay, @ostannard, @eugenis, paging you since you implemented or reviewed the original `--debug-vars` option, which this is closely tied to.

gulfemsavrun committed:
[llvm-objdump] Address reviewer feedback

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
+};
+
+class InlinedFunction : public LiveElement {
+  DWARFDie InlinedFuncDie;
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
+  virtual ~LiveElement() {};
+  const char *getName() const { return Name; }
+  DWARFUnit *getDwarfUnit() const { return Unit; }
+  const DWARFDie getFuncDie() const { return FuncDie; }
```
I removed `getDwarfUnit()` and `getFuncDie()`, but I kept `getName()` by `LiveElementPrinter` (originally LiveVariablePrinter) class methods.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
-
-  void print(raw_ostream &OS, const MCRegisterInfo &MRI) const;
+  virtual ~LiveVariable() {}
+  DWARFLocationExpression getLocationExpression() const { return LocExpr; };
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.cpp:
```diff
@@ -66,30 +147,37 @@ void LiveVariablePrinter::addVariable(DWARFDie FuncDie, DWARFDie VarDie) {
-        Child.getTag() == dwarf::DW_TAG_formal_parameter)
+    if (DbgVariables != DVDisabled &&
+        (Child.getTag() == dwarf::DW_TAG_variable ||
+         Child.getTag() == dwarf::DW_TAG_formal_parameter))
```
Done.

gulfemsavrun commented on llvm/tools/llvm-objdump/ObjdumpOpts.td:
```diff
@@ -248,7 +248,13 @@ def debug_vars_EQ : Joined<["--"], "debug-vars=">,
+def : Flag<["--"], "debug-inlined-funcs">, Alias<debug_inlined_funcs_EQ>, AliasArgs<["unicode"]>;
+
+def debug_indent_EQ : Joined<["--"], "debug-indent=">,
   HelpText<"Distance to indent the source-level variable display, "
```
I reordered the options. What kind of tweaking do you have in mind?

gulfemsavrun replied:
Ok, you are referring to `debug-indent` option, and I thought that you were referring to `debug-inlined-funcs` option. I'm going to address this. 

I also need to rebase my branch to resolve the conflicts. Can I use force push for that? Is it going to make it difficult for you to review?

gulfemsavrun committed:
[llvm-objdump] Address further feedback

1) Rename line format option to limits-only
2) Update debug-indent help text
3) Update debug-inlined-function.s test to check for demangling

gulfemsavrun committed:
Merge remote-tracking branch 'origin' into llvm-objdump

gulfemsavrun commented on llvm/tools/llvm-objdump/SourcePrinter.h:
```diff
@@ -22,60 +23,108 @@
-/// Helper class for printing source variable locations alongside disassembly.
+/// Helper class for printing source locations for variables and inlines
+/// subroutines alongside disassembly.
 class LiveVariablePrinter {
```
Ack.

jh7370 commented on llvm/test/tools/llvm-objdump/X86/debug-inlined-functions.s:
```diff
@@ -1,84 +1,104 @@
+# UNICODE-NEXT:      16: 01 f0                        	addl	%esi, %eax                                  ┻
+# UNICODE-NEXT:      18: c3                           	retq
+
+# MANGLE: 0000000000000000 <_Z3barii>:
```
You don't need an entire new set of CHECKS that are largely identical here. Instead, use multiple CHECK-PREFIXES per test set - one to indicate the mangling state and one to indicate the expected format. You'll need to use the --check-prefixes option instead. There are plenty of good examples in other tests, but basically, something like this would work, I think:

```
# ... | FileCheck --check-prefixes=UNICODE,UNICODE-MANGLED
# ... | FileCheck --check-prefixes=UNICODE,UNICODE-DEMANGLED

# UNICODE-MANGLED: 0000000000000000 <_Z3barii>:
# UNICODE-DEMANGLED: 0000000000000000 <bar(int, int)>:
# UNICODE-NEXT: ...
```

jh7370 commented on llvm/tools/llvm-objdump/ObjdumpOpts.td:
```diff
@@ -241,22 +241,22 @@ defm prefix_strip
 def debug_inlined_funcs_EQ : Joined<["--"], "debug-inlined-funcs=">,
   HelpText<"Print the locations of inlined functions alongside disassembly. "
-           "Supported formats: ascii, unicode (default) and line">,
+           "Supported formats: ascii, line and unicode (default)">,
```
```suggestion
           "Supported formats: ascii, line, and unicode (default)">,
```
Nit

jh7370 commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -155,15 +148,22 @@ OPTIONS
 
   Print the locations of inlined functions alongside disassembly.
-  ``format`` may be ``unicode``, ``ascii`` or ``line``, defaulting to
+  ``format`` may be ``ascii``, ``limits-only`` or ``unicode``, defaulting to
```
```suggestion
  ``format`` may be ``ascii``, ``limits-only``, or ``unicode``, defaulting to
```
Nit

gulfemsavrun replied:
Done.

gulfemsavrun committed:
[llvm-objdump] Simplify the test

gulfemsavrun committed:
[llvm-objdump] Fix a typo

jh7370 reviewed (approved):
LGTM! Do you need help merging or do you have commit access?

It might be worth leaving it a few days to land in case any of the reviewers (@MaskRay/ @echristo/@ostannard/@eugenis have anything to add).

gulfemsavrun committed:
Merge remote-tracking branch 'origin' into llvm-objdump

gulfemsavrun committed:
Merge remote-tracking branch 'origin' into llvm-objdump
